tests/Compat/FormDecorator/LabelDecoratorTest.php

```diff
@@ -23,4 +28,74 @@ public function testWithoutLabelAttribute(): void
 
         $this->assertSame('&nbsp;', $results->assemble()->render());
     }
+
+    public function testRequiredElement(): void
+    {
+        $formElement = new TextElement('test', [
+            'required' => true,
+            'label' => 'test-label'
+        ]);
+
+        $results = new FormElementDecorationResult();
+        $this->decorator->decorateFormElement($results, $formElement);
+
+        $this->assertStringContainsString(
+            '<span class="required-hint" aria-hidden="true" title="Required"> *</span>',
+            $results->assemble()->render()
+        );
+    }
+
+    public function testNonRequiredElement(): void
+    {
+        $formElement = new TextElement('test', [
+            'required' => false,
+            'label' => 'test-label'
+        ]);
+
+        $results = new FormElementDecorationResult();
+        $this->decorator->decorateFormElement($results, $formElement);
+
+        $this->assertStringNotContainsString(
+            '<span class="required-hint" aria-hidden="true" title="Required"> *</span>',
+            $results->assemble()->render()
+        );
+    }
+
+    public function testFormWithRequiredElement(): void
+    {
+        $formElement = new TextElement('test', [
+            'required' => true,
+            'label' => 'test-label'
+        ]);
+
+        $results = new FormElementDecorationResult();
+        $formStub = $this->createStub(Form::class);
+
+        $this->decorator->decorateFormElement($results, $formElement);
+        $this->decorator->decorateForm($results, $formStub);
+
+        $this->assertStringEndsWith(
+            '<ul class="form-info"><li>* Required field</li></ul>',
+            $results->assemble()->render()
+        );
+    }
+
+    public function testFormWithoutRequiredElements(): void
+    {
+        $formElement = new TextElement('test', [
+            'required' => false,
+            'label' => 'test-label'
+        ]);
+
+        $results = new FormElementDecorationResult();
+        $formStub = $this->createStub(Form::class);
+
+        $this->decorator->decorateFormElement($results, $formElement);
+        $this->decorator->decorateForm($results, $formStub);
+
+        $this->assertStringNotContainsString(
```

Comment:
As this represents the counterpart of `testFormWithRequiredElement`, what about using `assertStringEndsNotWith`? :P